lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1595,6 +1651,195 @@ class BuildBBRegs : public FuncOpPartialLoweringPattern {
+        // Instead, we wait until after BuildOpGroups, and replace all uses
+        // inside BuildPipelineGroups, once the pipeline register created here
+        // has been assigned to.
+      }
+    });
+    return success();
+  }
+};
+
+/// Builds groups for assigning registers for pipeline stages.
+class BuildPipelineGroups : public FuncOpPartialLoweringPattern {
+  using FuncOpPartialLoweringPattern::FuncOpPartialLoweringPattern;
+
+  LogicalResult
+  PartiallyLowerFuncToComp(mlir::FuncOp funcOp,
+                           PatternRewriter &rewriter) const override {
+    for (auto pipeline : funcOp.getOps<staticlogic::PipelineWhileOp>())
+      for (auto stage :
+           pipeline.getStagesBlock().getOps<staticlogic::PipelineStageOp>())
+        if (failed(buildStageGroups(pipeline, stage, rewriter)))
+          return failure();
+
+    return success();
+  }
+
+  LogicalResult buildStageGroups(staticlogic::PipelineWhileOp whileOp,
+                                 staticlogic::PipelineStageOp stage,
+                                 PatternRewriter &rewriter) const {
+    // Collect pipeline registers for stage.
+    auto pipelineRegisters = getComponentState().getPipelineRegs(stage);
+
+    // Collect group names for the prologue or epilogue.
+    SmallVector<StringAttr> prologueGroups;
+    SmallVector<StringAttr> epilogueGroups;
+
+    // Get the number of pipeline stages, excluding the termina
+    size_t numStages = whileOp.getStagesBlock().getOperations().size() - 1;
+
+    // For each registered stage result value.
+    for (auto &operand :
+         stage.getBodyBlock().getTerminator()->getOpOperands()) {
+      unsigned i = operand.getOperandNumber();
+      Value value = operand.get();
+
+      // Get the pipeline register for that result.
+      auto pipelineRegister = pipelineRegisters[i];
+
+      // Get the evaluating group for that value.
+      auto evaluatingGroup = getComponentState().getEvaluatingGroup(value);
+
+      // Remember the final group for this stage result.
+      calyx::GroupOp group;
+
+      // Stitch the register in, depending on whether the group was
+      // combinational or sequential.
+      if (auto combGroup =
+              dyn_cast<calyx::CombGroupOp>(evaluatingGroup.getOperation()))
+        group =
+            convertCombToSeqGroup(combGroup, pipelineRegister, value, rewriter);
+      else
+        group =
+            replaceGroupRegister(evaluatingGroup, pipelineRegister, rewriter);
+
+      // Replace the stage result uses with the register out.
+      stage.getResult(i).replaceAllUsesWith(pipelineRegister.out());
+
+      // Mark the group for scheduling in the pipeline's block.
+      getComponentState().addBlockScheduleable(stage->getBlock(), group);
+
+      // Add the group to the prologue or epilogue as necessary.
+      unsigned stageNumber = stage.getStageNumber();
+      if (stageNumber < numStages - 1)
```

Comment:
1. Why is this `-1`?
2. Can we avoid the subtraction in unsigned, or add an assert?

Reply:
This is to say "all stages but the last". I will add a better comment here. Added an assertion that `numStages` is greater than 0 so this subtraction will never go "negative".